### ISSUE TITLE
FileCompiler doesn't compile contents of constant()

### ISSUE THREAD

**Issue by teppokoivula** [processwire/processwire-issues#179]
### Short description of the issue

In a module I'm using constant() to check if a specific constant is defined, and to return the value of said constant. Name of the constant is dynamically defined, which is why I can't use hard-coded values here.

It appears that FileCompiler doesn't understand this at all: `constant("Template::flagSystem")`. It remains as-is in the compiled file, even though it *should* become `constant("\ProcessWire\Template::flagSystem")`.

In this case it doesn't (and shouldn't) matter whether the value is dynamically generated or hard-coded, neither option works (tested). That being said, in my use case constant gets actually called with `constant("Template::flag{$value}")`.

### Setup/Environment

* ProcessWire version: 3.0.51
* (Optional) PHP version: 5.5.21
* (Optional) Any 3rd party modules that are installed and could be related to the issue: ProcessChangelog

**Comment by ryancramerdesign** [comment]
Other than a couple of rare cases, FileCompiler doesn't go hunting around in strings for classes. There's just too much potential for false positives once you are in PHP strings. If you can, I'd suggest referencing your class name outside of the string, like this:
~~~~~~
constant(Template::class . "::flag{$value}");
~~~~~~

**Comment by teppokoivula** [comment]
The ::class constant is not really an option, as it requires PHP 5.5. Thanks for letting me know why this is happening, though :)

In another module I have implemented a (only slightly hacky) solution that makes use of ReflectionClass to work around some 2.x/3.x incompatibilities, I guess I cold use that here too. Another option would be using a hardcoded values instead.

Closing this for now. Not sure if such a thing already exists, but it would be great to have an easy to ise list of FileCompiler do's and dont's so that one doesn't have to figure them out the hard way :)
